src/Orleans.Core/Timers/TimerManager.cs

```diff
@@ -0,0 +1,363 @@
+        {
+            return threadLocalQueue ?? (threadLocalQueue = InitializeThreadLocalQueue());
+
+            ThreadLocalQueue InitializeThreadLocalQueue()
+            {
+                var threadLocals = new ThreadLocalQueue();
+                while (true)
+                {
+                    lock (AllQueuesLock)
+                    {
+                        var queues = Volatile.Read(ref allQueues);
+
+                        // Find a spot in the existing array to register this thread.
+                        for (var i = 0; i < queues.Length; i++)
+                        {
+                            if (Volatile.Read(ref queues[i]) == null)
+                            {
+                                Volatile.Write(ref queues[i], threadLocals);
+                                return threadLocals;
+                            }
+                        }
+
+                        // The existing array is full, so copy all values to a new, larger array and register this thread.
+                        var newQueues = new ThreadLocalQueue[queues.Length * 2];
+                        Array.Copy(queues, newQueues, queues.Length);
+                        newQueues[queues.Length] = threadLocals;
+                        Volatile.Write(ref allQueues, newQueues);
+                        return threadLocals;
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Holds per-thread timer data.
+        /// </summary>
+        private sealed class ThreadLocalQueue : ILinkedList<T>
+        {
+            public readonly RecursiveInterlockedExchangeLock Lock = new RecursiveInterlockedExchangeLock();
+
+            /// <summary>
+            /// The number of times that this queue has been starved since it was last serviced.
+            /// </summary>
+            public int StarvationCount;
+
+            public T Head { get; set; }
+
+            public T Tail { get; set; }
+        }
+
+        /// <summary>
+        /// Holds timers that have expired and should be fired.
+        /// </summary>
+        private sealed class ExpiredTimers : ILinkedList<T>
+        {
+            public T Head { get; set; }
+
+            public T Tail { get; set; }
+
+            public void FireTimers()
+            {
+                var current = this.Head;
+                while (current != null)
+                {
+                    try
+                    {
+                        current.OnTimeout();
+                    }
+                    catch
+                    {
+                        // Ignore any exceptions during firing.
+                    }
+
+                    current = current.Next;
+                }
+            }
+        }
+    }
+
+    internal interface ITimerCallback
+    {
+        /// <summary>
+        /// The UTC time when this timer is due.
+        /// </summary>
+        DateTime DueTime { get; }
+
+        bool IsCanceled { get; }
```

Comment:
`IsCanceled` doesn't appear to be used.

Reply:
In effect, it's not. It's from Dmitry's code. The reason for it was for things such as request timeouts, where usually we cancel the timer before it fires. It could also serve to support `CancellationToken`.

I can remove it for now, though, if that's preferred.

Reply:
That's fine with me.

Reply:
I removed it. We will have the PR branch to consult if we ever want to bring it back